casr/src/util.rs

```diff
@@ -293,3 +282,125 @@ pub fn get_output(command: &mut Command, timeout: u64, error_on_timeout: bool) -
+///
+/// # Arguments
+///
+/// * `matches` - tool arguments
+///
+/// # Return value
+///
+/// Path to output directory
+pub fn initialize_dirs(matches: &clap::ArgMatches) -> Result<&PathBuf> {
+    // Get output dir
+    let output_dir = matches.get_one::<PathBuf>("output").unwrap();
+    if !output_dir.exists() {
+        fs::create_dir_all(output_dir).with_context(|| {
+            format!("Couldn't create output directory {}", output_dir.display())
+        })?;
+    } else if output_dir.read_dir()?.next().is_some() {
+        bail!("Output directory is not empty.");
+    }
+    // Get oom dir
+    let oom_dir = output_dir.join("oom");
+    if fs::create_dir_all(&oom_dir).is_err() {
+        bail!("Failed to create dir {}", &oom_dir.to_str().unwrap());
+    }
+    // Get timeout dir
+    let timeout_dir = output_dir.join("timeout");
+    if fs::create_dir_all(&timeout_dir).is_err() {
+        bail!("Failed to create dir {}", &timeout_dir.to_str().unwrap());
+    }
+
+    Ok(output_dir)
+}
+
+/// Method checks whether binary file contains predefined symbols.
+///
+/// # Arguments
+///
+/// * `path` - path to binary to check.
+///
+/// # Return value
+///
+/// Set of important symbols
+pub fn symbols_list(path: &Path) -> Result<HashSet<&str>> {
+    let mut found_symbols = HashSet::new();
+    if let Ok(buffer) = fs::read(path) {
+        if let Ok(elf) = goblin::elf::Elf::parse(&buffer) {
+            let symbols = [
+                "__asan",
+                "__ubsan",
+                "__tsan",
+                "__msan",
+                "__llvm_profile",
+                "runtime.go",
+            ];
+            for sym in elf.syms.iter() {
+                if let Some(name) = elf.strtab.get_at(sym.st_name) {
+                    for symbol in symbols.iter() {
+                        if name.contains(symbol) {
+                            found_symbols.insert(*symbol);
+                            break;
+                        }
+                    }
+                }
+            }
+        } else {
+            bail!("Fuzz target: {} must be an ELF executable.", path.display());
+        }
+    } else {
+        bail!("Couldn't read fuzz target binary: {}.", path.display());
+    }
+
+    Ok(found_symbols)
+}
+
+/// Function searches for path to the tool
+///
+/// # Arguments
+///
+/// * 'tool' - tool name
+///
+/// # Return value
+///
+/// Path to the tool
+pub fn get_path(tool: &str) -> Result<PathBuf> {
+    let mut path_to_tool = std::env::current_exe()?;
+    let current_tool = path_to_tool
+        .file_name()
+        .unwrap()
+        .to_str()
+        .unwrap()
+        .to_string();
+    path_to_tool.pop();
+    path_to_tool.push(tool);
+    if path_to_tool.is_executable() {
+        Ok(path_to_tool)
+    } else if let Ok(path_to_tool) = which::which(tool) {
+        Ok(path_to_tool)
+    } else {
+        bail!(
+            "{path_to_tool:?}: No {tool} next to {current_tool}. And there is no {tool} in PATH."
```

Comment:
Are you sure that bail supports string formatting? Also, check similar places, please

Reply:
yes, bail supports string formating